lib/generators/deimos/schema_class_generator.rb

```diff
@@ -0,0 +1,231 @@
+# frozen_string_literal: true
+
+require 'rails/generators'
+require 'deimos'
+require 'deimos/utils/schema_class_mixin'
+require 'deimos/config/configuration'
+
+# Generates new schema classes.
+module Deimos
+  module Generators
+    # Generator for Schema Classes used for the IDE and consumer/producer interfaces
+    class SchemaClassGenerator < Rails::Generators::Base
+      include Deimos::Utils::SchemaClassMixin
+
+      SPECIAL_TYPES = %i(record enum).freeze
+
+      source_root File.expand_path('schema_class/templates', __dir__)
+
+      argument :full_schema,
+               desc: 'The fully qualified schema name or path.',
+               required: false,
+               type: :string
+
+      no_commands do
+        # Retrieve the fields from this Avro Schema
+        # @return [Array<SchemaField>]
+        def fields
+          @current_schema.fields.map { |field| Deimos::SchemaField.new(field.name, field.type) }
+        end
+
+        # Converts Deimos::SchemaField's to String form for generated YARD docs
+        # @param schema_field [Deimos::SchemaField]
+        # @return [String] A string representation of the Type of this SchemaField
+        def deimos_field_type(schema_field)
+          _field_type(schema_field.type)
+        end
+
+        # @param enum [Avro::Schema::EnumSchema] a field of type 'enum'.
+        # @return [Array<String>] of symbols valid for the enum.
+        def enum_symbols(enum)
+          enum.symbols
+        end
+
+        # @param enum [Avro::Schema::EnumSchema] a field of type 'enum'.
+        # @return [String] the possible return values for this Enum type
+        def enum_return_values(enum)
+          "'#{enum.symbols.join("', '")}'"
+        end
+
+        # Generate a Schema Model Class and all of its Nested Records from an Avro Schema
+        # @param schema_name [String] the name of the Avro Schema in Dot Syntax
+        def generate_classes_from_schema(schema_name)
+          _schema_base(schema_name).load_schema!
+          _schema_base.schema_store.schemas.each_value do |schema|
+            @current_schema = schema
+            @special_field_initialization = schema.type_sym == :record ? special_field_initialization : {}
+            file_prefix = schema.name.underscore
+            namespace_path = schema.namespace.tr('.', '/')
+            schema_template = "schema_#{schema.type}.rb"
+            filename = "#{Deimos.config.schema.generated_class_path}/#{namespace_path}/#{file_prefix}.rb"
+            template(schema_template, filename, force: true)
+          end
+          _clear_schema_base!
+        end
+
+        # Retrieve any special formatting needed for this current schema's fields
+        # @return [Hash<String, Array[Symbol]>]
+        def special_field_initialization
+          result = Hash.new { |h, k| h[k] = [] }
+          fields.each do |field|
+            field_base_type = field.type.type_sym # Record, Union, Enum, Array or Map?
```

Comment:
Yeah there's a lot of Avro-specific stuff here. Let's see if we can move some of it into the backend base. Hopefully it can be as simple as "here's a schema, here's a field, is the Ruby object an array, an object, an enum or a scalar value?"

Reply:
Removed this function ;)